src/main/java/com/mraof/minestuck/entity/AnimatedCreatureEntity.java

```diff
@@ -0,0 +1,165 @@
+package com.mraof.minestuck.entity;
+
+import net.minecraft.entity.CreatureEntity;
+import net.minecraft.entity.EntityType;
+import net.minecraft.entity.LivingEntity;
+import net.minecraft.entity.ai.goal.MeleeAttackGoal;
+import net.minecraft.network.datasync.DataParameter;
+import net.minecraft.network.datasync.DataSerializers;
+import net.minecraft.network.datasync.EntityDataManager;
+import net.minecraft.world.World;
+import software.bernie.geckolib3.core.IAnimatable;
+import software.bernie.geckolib3.core.controller.AnimationController;
+import software.bernie.geckolib3.core.manager.AnimationFactory;
+
+public abstract class AnimatedCreatureEntity extends CreatureEntity implements IAnimatable {
+    private final AnimationFactory factory = new AnimationFactory(this);
+    protected static final DataParameter<Integer> CURRENT_ACTION = EntityDataManager.defineId(AnimatedCreatureEntity.class, DataSerializers.INT);
+
+    private int heavyAttackTicks = 0;
+    private int recoveryTicks = 0;
+    private int attackDelay;
+    private int attackRecovery;
+    private int actionResetTicks;
+
+    protected boolean canMoveWhileAttacking = false;
+
+    protected AnimatedCreatureEntity(EntityType<? extends CreatureEntity> type, World world) {
+        super(type, world);
+    }
+
+    @Override
+    protected void registerGoals() {
+        super.registerGoals();
+        goalSelector.addGoal(3, new DelayedAttackGoal(this, 1F, false));
+    }
+
+    @Override
+    protected void defineSynchedData() {
+        super.defineSynchedData();
+        entityData.define(CURRENT_ACTION, Actions.NONE.ordinal());
+    }
+
+    @Override
+    public void tick() {
+        super.tick();
+        if (heavyAttackTicks > 0) {
+            heavyAttackTicks--;
+            if (heavyAttackTicks == 0) {
+                recoveryTicks = attackRecovery;
+                if (getTarget() != null && isInRange(getTarget())) { // TODO: AOE bounding box collision checks + aoe flag
+                    doHurtTarget(getTarget());
+                }
+            }
+        }
+        if (recoveryTicks > 0) {
+            recoveryTicks--;
+            if (recoveryTicks == 0) {
+                setAttacking(false);
+            }
+        }
+        if (getCurrentAction() != Actions.NONE && actionResetTicks > 0) {
+            actionResetTicks--;
+            if (actionResetTicks <= 0) {
+                setCurrentAction(Actions.NONE, 0);
+            }
+        }
```

Comment:
Is there any overlap in the way recoveryTicks and actionResetTicks are used?

Reply:
yeah oops, i need to rework that. i used that before i put the attack in actions. it should never overlap